### PR TITLE
Add the exception_ptr of the unhandled exception to retry_context

### PR DESCRIPTION
**Pull request by rahuldutta90**
Context:
This will give a better handling of exceptions especially winhttp errors (https://docs.microsoft.com/en-us/windows/win32/winhttp/error-messages) from clients overriding the azure::storage::Retry_policy. Currently our service is overriding your retry_policy. This gives us more flexibility to handle retries on append scenarios.

1) Added exception_ptr to retry_context
2) Pass the exception other than storage_exception to retry_context in excecutor.

### CODE REVIEW

File: Microsoft.WindowsAzure.Storage/includes/was/core.h [hunk 1]
```diff
@@ -1275,12 +1277,21 @@ namespace azure { namespace storage {
             return m_current_location_mode;
         }
 
+        /// <summary>
+        /// Gets the location mode for subsequent retries.
+        /// </summary>
+        /// <returns>The <see cref="azure::storage::location_mode" /> for subsequent retries.</returns>
+        const std::exception_ptr& unhandled_exception() const {
```
Comment (katmsft) [inline Microsoft.WindowsAzure.Storage/includes/was/core.h]: The added documentation is patching to the wrong place, please review it.

File: Microsoft.WindowsAzure.Storage/src/executor.cpp [hunk 2]
```diff
@@ -297,6 +298,10 @@ namespace azure { namespace storage { namespace core {
                         retryable_exception = e.retryable();
                         throw;
```
Comment (katmsft) [inline Microsoft.WindowsAzure.Storage/src/executor.cpp]: If the `ex_ptr` is not set here, I believe all `storage_exception` will not be available when retrieving unhandled exception.

File: Microsoft.WindowsAzure.Storage/includes/was/core.h [hunk 3]
```diff
@@ -1275,12 +1277,21 @@ namespace azure { namespace storage {
...
+            return m_unhandled_exception;
+        }
+
     private:
 
         int m_current_retry_count;
         request_result m_last_request_result;
         storage_location m_next_location;
         location_mode m_current_location_mode;
+        std::exception_ptr m_unhandled_exception;
```
Comment (rahuldutta90) [inline Microsoft.WindowsAzure.Storage/includes/was/core.h]: change name and doc of m_unhandled_exception and its getter

File: Microsoft.WindowsAzure.Storage/src/executor.cpp (hunk 2 above)
Comment (rahuldutta90) [reply Microsoft.WindowsAzure.Storage/src/executor.cpp]: discussed this offline yesterday, for now storing nonstorage exceptions

**Comment by JinmingHu-MSFT** [comment]
Unit test looks good

**Comment by rahuldutta90** [comment]
@katmsft  I see tests failing. I checked out master and there also see these test failing. Shouldnt be related to my changes right?

**Comment by JinmingHu-MSFT** [comment]
@rahuldutta90 Don't worry. That's the problem of test cases.

**Pull request merged**